File: your.app.folder/scene_manager.py
"""
Scene Manager
Handles scene transitions and state management
"""


class SceneManager:
    def __init__(self):
        self.scenes = {}
        self.current_scene = None
        self.current_scene_name = None
    
    def register_scene(self, name, scene):
        """Register a scene with a name"""
        self.scenes[name] = scene
    
    def change_scene(self, name):
        """Change to a different scene"""
        if name in self.scenes:
            # Call on_exit on current scene if it exists
            if self.current_scene:
                if hasattr(self.current_scene, 'on_exit'):
                    self.current_scene.on_exit()
            
            # Switch to new scene
            self.current_scene_name = name
            self.current_scene = self.scenes[name]
            
            # Call on_enter on new scene
            if hasattr(self.current_scene, 'on_enter'):
                self.current_scene.on_enter()
        else:
            print(f"Warning: Scene '{name}' not found")
    
    def handle_event(self, event):
        """Pass event to current scene"""
        if self.current_scene:
            self.current_scene.handle_event(event)
    
    def update(self):
        """Update current scene"""
        if self.current_scene:
            self.current_scene.update()
    
    def render(self, screen):
        """Render current scene"""
        if self.current_scene:
            self.current_scene.render(screen)

File: src/scene_manager.ts
/**
 * Scene Manager
 * Handles scene transitions and state management
 */

export interface Scene {
  onEnter?(): void;
  onExit?(): void;
  handleEvent?(event: KeyboardEvent | MouseEvent): void;
  update?(deltaTime: number): void;
  render?(ctx: CanvasRenderingContext2D): void;
}

export class SceneManager {
  private scenes: Map<string, Scene> = new Map();
  private currentScene: Scene | null = null;
  private currentSceneName: string | null = null;

  registerScene(name: string, scene: Scene): void {
    this.scenes.set(name, scene);
  }

  changeScene(name: string): void {
    if (this.scenes.has(name)) {
      // Call on_exit on current scene if it exists
      if (this.currentScene?.onExit) {
        this.currentScene.onExit();
      }

      // Switch to new scene
      this.currentSceneName = name;
      this.currentScene = this.scenes.get(name)!;

      // Call on_enter on new scene
      if (this.currentScene?.onEnter) {
        this.currentScene.onEnter();
      }
    } else {
      console.warn(`Warning: Scene '${name}' not found`);
    }
  }

  handleEvent(event: KeyboardEvent | MouseEvent): void {
    if (this.currentScene?.handleEvent) {
      this.currentScene.handleEvent(event);
    }
  }

  update(deltaTime: number): void {
    if (this.currentScene?.update) {
      this.currentScene.update(deltaTime);
    }
  }

  render(ctx: CanvasRenderingContext2D): void {
    if (this.currentScene?.render) {
      this.currentScene.render(ctx);
    }
  }

  getCurrentSceneName(): string | null {
    return this.currentSceneName;
  }
}

